src/Features/Core/Portable/UnusedReferences/ProjectAssets/ProjectAssetsReader.cs

```diff
@@ -146,23 +138,43 @@ internal static class ProjectAssetsReader
 
                 if (targetLibrary.Compile != null)
                 {
-                    compilationAssemblies.AddRange(targetLibrary.Compile.Keys
-                        .Where(assemblyPath => !assemblyPath.EndsWith(NuGetEmptyFileName))
-                        .Select(assemblyPath => Path.GetFullPath(Path.Combine(packagesPath, library.Path, assemblyPath))));
+                    foreach (var kvp in targetLibrary.Compile)
+                    {
+                        var assemblyPath = kvp.Key;
+                        if (!assemblyPath.EndsWith(NuGetEmptyFileName))
+                        {
+                            compilationAssemblies.Add(Path.GetFullPath(Path.Combine(packagesPath, library.Path ?? "", assemblyPath)));
+                        }
+                    }
                 }
             }
 
             if (referenceType == ReferenceType.Unknown)
             {
                 return null;
             }
+            
+            if (referenceType == ReferenceType.Package && itemSpecification == ".NETStandard.Library")
+            {
+                // This depenedency is large and not useful in determining whether the parent reference has been used.
+                return null;
+            }
+
+            if (builtReferences.TryGetValue(itemSpecification, out var builtReference))
+            {
+                return builtReference;
+            }
 
             var dependencies = dependencyNames
-                .Select(dependency => BuildReference(projectAssets, dependency, treatAsUsed: false))
+                .Select(dependency => BuildReference(projectAssets, dependency, treatAsUsed: false, targetLibraryKeys, builtReferences))
                 .WhereNotNull()
                 .ToImmutableArray();
 
-            return new ReferenceInfo(referenceType, itemSpecification, treatAsUsed, compilationAssemblies.ToImmutable(), dependencies);
+            var reference = new ReferenceInfo(referenceType, itemSpecification, treatAsUsed, compilationAssemblies.ToImmutable(), dependencies);
+
+            builtReferences.Add(reference.ItemSpecification, reference);
+
+            return reference;
```

Comment:
this is a bit of a strange API in that it reference both the reference, but also adds it to a dictionary.  i do not like that at all :D   can you just return the reference and have someone higher up add to the dictionary?

Reply:
The issue with moving it is that this method is recursively invoked when building the dependencies list.